tokio-postgres/src/copy_both.rs

```diff
@@ -0,0 +1,350 @@
+/// matter what the users do with their CopyBothDuplex handle we're always going to send the
+/// correct messages to the backend in order to restore the connection into a usable state.
+///
+/// ```ignore
+///                                          |
+///          <tokio_postgres owned>          |    <userland owned>
+///                                          |
+///  pg -> Connection -> CopyBothReceiver ---+---> CopyBothDuplex
+///                                          |          ^   \
+///                                          |         /     v
+///                                          |      Sink    Stream
+/// ```
+pub struct CopyBothReceiver {
+    /// Receiver of backend messages from the underlying [Connection](crate::Connection)
+    message_receiver: mpsc::Receiver<Result<Message, Error>>,
+    /// Receiver of frontend messages sent by the user using <CopyBothDuplex as Sink>
+    sink_receiver: mpsc::Receiver<FrontendMessage>,
+    /// Sender of CopyData contents to be consumed by the user using <CopyBothDuplex as Stream>
+    stream_sender: mpsc::Sender<Result<Bytes, Error>>,
+    /// The current state of the subprotocol
+    state: CopyBothState,
+    /// Holds a buffered message until we are ready to send it to the user's stream
+    buffered_message: Option<Result<Bytes, Error>>,
+}
+
+impl CopyBothReceiver {
+    pub(crate) fn new(
+        message_receiver: mpsc::Receiver<Result<Message, Error>>,
+        sink_receiver: mpsc::Receiver<FrontendMessage>,
+        stream_sender: mpsc::Sender<Result<Bytes, Error>>,
+    ) -> CopyBothReceiver {
+        CopyBothReceiver {
+            message_receiver,
+            sink_receiver,
+            stream_sender,
+            state: CopyBothState::CopyBoth,
+            buffered_message: None,
+        }
+    }
+
+    /// Convenience method to set the subprotocol into an unexpected message state
+    fn unexpected_message(&mut self) {
+        self.sink_receiver.close();
+        self.buffered_message = Some(Err(Error::unexpected_message()));
+        self.state = CopyBothState::CommandComplete;
+    }
+
+    /// Processes messages from the backend, it will resolve once all backend messages have been
+    /// processed
+    fn poll_backend(&mut self, cx: &mut Context<'_>) -> Poll<()> {
+        use CopyBothState::*;
+
+        loop {
+            // Deliver the buffered message (if any) to the user to ensure we can potentially
+            // buffer a new one in response to a server message
+            if self.buffered_message.is_some() {
+                // If the receiver is gone we'll just drop the message
+                let _ = ready!(self.stream_sender.poll_ready(cx));
+                let message = self
+                    .buffered_message
+                    .take()
+                    .expect("already checked in outer if");
+                let _ = self.stream_sender.start_send(message);
+            }
```

Comment:
What I was getting at before was just a style issue, but I believe the following is semantically equivalent and more concise:

```
            if let Some(message) = self.buffered_message.take() {
                // If the receiver is gone we'll just drop the message                                                    
                let _ = ready!(self.stream_sender.poll_ready(cx));
                let _ = self.stream_sender.start_send(message);
            }
```

But I could have made a mistake, so please check that it really does mean the same thing.

Reply:
I see. This won't work because `ready!()` desugars to:

```rust
            if let Some(message) = self.buffered_message.take() {
                // If the receiver is gone we'll just drop the message                                                    
                let _ = match self.stream_sender.poll_ready(cx) {
                    Poll::Ready(val) => val,
                    Poll::Pending => return Poll::Pending, // XXX message lost in this case
                };
                let _ = self.stream_sender.start_send(message);
            }
```
Which means that if the user is consuming their stream slowly and we get `Pending` back, the value will be dropped because `take()` has already been called.

Reply:
What could work, but has wonky semantics imo is this:

```rust
            if self.stream_sender.poll_ready(cx).is_ready() {
                if let Some(message) = self.buffered_message.take() {
                    let _ = self.stream_sender.start_send(message);
                }
            }
```

This feels strange to me because technically we're waiting on a stream handle to become ready but then we might not use it

Reply:
I pushed a version that I think ticks all the boxes :)